### PR TITLE
chore: add initial oauth authentication wrapper to call GAF

### PR DESCRIPTION
**Pull request by bastiandoetsch**
### Description

Not used yet, but adds the functionality to call GAF via a language server auth provider.

### Checklist

- [x] Tests added and all succeed
- [x] Linted
- [ ] README.md updated, if user-facing
- [x] License file updated, if new 3rd-party dependency is introduced

### Screenshots / GIFs

_Visuals that may help the reviewer. Please add screenshots for any UI change. GIFs are most welcome!_

### CODE REVIEW

File: application/config/config.go [hunk 1]
```diff
@@ -339,6 +344,9 @@ func (c *Config) UpdateApiEndpoints(snykApiUrl string) bool {
 		}
 
 		c.setSnykCodeApi(snykCodeApiUrl)
+		if strings.Contains(snykCodeApiUrl, "fedramp") {
```
Comment (PeterSchafer) [inline application/config/config.go:347]: probably better snykgov.io

File: application/config/config_test.go [hunk 2]
```diff
@@ -168,6 +169,13 @@ func TestSnykCodeApi(t *testing.T) {
 		codeApiEndpoint, _ := getCodeApiUrlFromCustomEndpoint("")
 		assert.Equal(t, customDeeproxyUrl, codeApiEndpoint)
 	})
+
+	t.Run("fedramp substring endpoint enables oauth authentication", func(t *testing.T) {
+		endpoint := "https://app.fedramp,snyk.io/api/v1"
```
Comment (PeterSchafer) [inline application/config/config_test.go:174]: snykgov.io and typo

File: application/di/init.go [hunk 3]
```diff
@@ -105,6 +109,7 @@ func initInfrastructure() {
 		},
 	)
 
+	engine = app.CreateAppEngine()
```
Comment (PeterSchafer) [inline application/di/init.go:112]: init missing and maybe re-use common method here and in TestInit()

File: application/server/configuration_test.go [hunk 4]
```diff
@@ -420,4 +420,16 @@ func Test_InitializeSettings(t *testing.T) {
 
 		assert.Equal(t, true, c.IsSnykCodeQualityEnabled())
 	})
+
+	t.Run("authenticationMethod is passed", func(t *testing.T) {
+		c := config.New()
+		config.SetCurrentConfig(c)
+		di.TestInit(t)
+		authenticationMethod := lsp.OAuthAuthentication
+
```
Comment (PeterSchafer) [inline application/server/configuration_test.go:429]: please check the pre-condition

File: infrastructure/oauth/oauth_provider.go [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+
+func (p *oAuthProvider) ClearAuthentication(ctx context.Context) error {
+	p.config.Set(auth.CONFIG_KEY_OAUTH_TOKEN, "")
+	p.config.Set(configuration.AUTHENTICATION_TOKEN, "")
+	p.config.Set(configuration.AUTHENTICATION_BEARER_TOKEN, "")
+	return nil
+}
+
+func (p *oAuthProvider) AuthURL(ctx context.Context) string {
+	// TODO: this is a hack, we should be able to get the auth url from the authenticator
```
Comment (PeterSchafer) [inline infrastructure/oauth/oauth_provider.go:59]: return p.auythURL

File: infrastructure/oauth/oauth_provider.go [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+	return tokenUrl
+}
+
+func (p *oAuthProvider) AuthenticateToken(token string) error {
+	var oauthToken oauth2.Token
+	err := json.Unmarshal([]byte(token), &oauthToken)
+	if err != nil {
+		return err
+	}
+	if oauthToken.Expiry.Before(time.Now()) {
```
Comment (PeterSchafer) [inline infrastructure/oauth/oauth_provider.go:70]: Please use https://pkg.go.dev/golang.org/x/oauth2#Token.Valid

File: infrastructure/oauth/oauth_provider_test.go [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+	config.Set(configuration.WEB_APP_URL, "https://app.fake.snyk.io")
+	authenticator := NewFakeOauthAuthenticator(time.Now().Add(10*time.Second), true, config).(*fakeOauthAuthenticator)
+	provider := NewOAuthProvider(config, authenticator)
+
+	url := provider.AuthURL(context.Background())
+
+	assert.NotEmpty(t, url)
+	_, err := url2.Parse(url)
+	assert.NoError(t, err)
+	assert.Equal(t, url, "https://id.fake.snyk.io/oauth2/default/v1/token")
```
Comment (PeterSchafer) [inline infrastructure/oauth/oauth_provider_test.go:170]: too specific, maybe just check that the url is parseable

File: application/config/config.go [hunk 8]
```diff
@@ -201,11 +203,12 @@ func New() *Config {
 	c.snykCodeAnalysisTimeout = snykCodeAnalysisTimeoutFromEnv()
 	c.token = ""
 	c.trustedFoldersFeatureEnabled = true
+	c.automaticScanning = true
```
Comment (PeterSchafer) [inline application/config/config.go:206]: how is `automaticScanning` related?

File: infrastructure/oss/maven_range_finder.go [hunk 9]
```diff
@@ -30,13 +29,11 @@ type mavenRangeFinder struct {
 }
 
 func (m *mavenRangeFinder) find(issue ossIssue) snyk.Range {
-	searchPackage, version := introducingPackageAndVersion(issue)
-	log.Trace().Interface("issue", issue).Str("searchPackage", searchPackage).Str("searchVersion", version)
```
Comment (PeterSchafer) [inline infrastructure/oss/maven_range_finder.go:34]: why was this removed? noise?

File: infrastructure/oss/oss.go [hunk 10]
```diff
@@ -238,10 +238,8 @@ func (oss *Scanner) unmarshallAndRetrieveAnalysis(ctx context.Context, res []byt
 		targetFileUri := uri.PathToUri(targetFilePath)
 		fileContent, err := os.ReadFile(targetFilePath)
 		if err != nil {
-			log.Err(err).Str("method", "unmarshallAndRetrieveAnalysis").
-				Msgf("Error while reading the file %v, err: %v", targetFile, err)
-			oss.errorReporter.CaptureErrorAndReportAsIssue(path, err)
-			return nil
+			// don't fail the scan if we can't read the file. No annotations with ranges, though.
+			fileContent = []byte{}
```
Comment (PeterSchafer) [inline infrastructure/oss/oss.go:242]: these changes seem unrelated. What is the reason for them?

File: application/config/config.go (hunk 8 above)
Comment (bastiandoetsch) [reply application/config/config.go:206]: it's not. was there before.

File: infrastructure/oss/oss.go (hunk 10 above)
Comment (bastiandoetsch) [reply infrastructure/oss/oss.go:242]: True. LS was erroring when not having a defined targetFile, e.g. when scanning unmanaged. this should make it a bit more resilient - instead of erroring out, it now just skips range finding.

Comment (PeterSchafer) [reply infrastructure/oss/oss.go:242]: do we have any tests that could cover this change?

Comment (bastiandoetsch) [reply infrastructure/oss/oss.go:242]: No yet, but I can write a test

File: application/server/configuration.go [hunk 11]
```diff
@@ -127,6 +132,29 @@ func writeSettings(settings lsp.Settings, initialize bool) {
...
+}
+
+func updateAuthenticationMethod(settings lsp.Settings) {
+	c := config.CurrentConfig()
+	c.SetAuthenticationMethod(settings.AuthenticationMethod)
+	if config.CurrentConfig().GetAuthenticationMethod() == lsp.OAuthAuthentication {
+		engine := di.Engine()
+		c := engine.GetConfiguration()
+		c.Set(configuration.OAUTH_AUTH_ENABLED, true)
+		c.Set(configuration.API_URL, config.CurrentConfig().SnykApi())
```
Comment (PeterSchafer) [inline application/server/configuration.go:145]: it might make sense to do this somewhere else, since we need this with and without oauth

Comment (bastiandoetsch) [reply application/server/configuration.go:145]: yep, will come with config sync

File: application/di/init.go [hunk 12]
```diff
@@ -93,27 +97,39 @@ func initDomain() {
...
+				"/usr/java",
+				"/opt",
+				"/Library",
+			})
+	}
 
+	initializeWorkflowEngine()
 	errorReporter = sentry2.NewSentryErrorReporter()
-	installer = install.NewInstaller(errorReporter)
+	installer = install.NewInstaller(errorReporter, engine.GetNetworkAccess().GetUnauthorizedHttpClient)
```
Comment (PeterSchafer) [inline application/di/init.go:121]: Just FYI, I think it would it would have also been fine to use the GetHttpClient() here.

File: internal/httpclient/httpclient.go [hunk 13]
```diff
@@ -17,42 +17,18 @@
...
-		log.Err(err).Str("method", method).Send()
-	}
-	if proxy != nil {
-		proxySplit := strings.Split(proxy.String(), "@")
-		proxyLogString := proxySplit[0]
-		if len(proxySplit) > 1 {
-			proxyLogString = "xxx@" + proxySplit[1]
-		}
-		log.Info().Str("method", method).Str("proxy", proxyLogString).Msg("created http client with proxy support")
+func NewHTTPClient(access networking.NetworkAccess) *http.Client {
```
Comment (PeterSchafer) [inline internal/httpclient/httpclient.go:25]: Question: these functions appear to be used only once. Shall we maybe just get rid of them?

Comment (bastiandoetsch) [reply internal/httpclient/httpclient.go:25]: let's do that in the next PR though. this one is big enough.
